---
title: "Introduction to single-cell analyses"
date: "`r Sys.Date()`"
author: "Velten group"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 1
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```


# Introduction

In this markdown we will be working with the data from the following study: [Multi-omics profiling of mouse gastrulation at single-cell resolution](https://www.nature.com/articles/s41586-019-1825-8). This study sequenced mouse embryonic stem cells as they undergo the process of gastrulation, during which the three primary germ layers are formed. In addition to the transcriptome, they also measured DNA methylation and chromatin accessibility.

# Load libraries
To get started please load the following libraries that you will need for this analysis.
```{r, message=FALSE}
library(data.table)  # fast manipulation of data.frames
library(purrr)       # pipes to make the code more readable
library(Seurat)
library(ggplot2)
library(MOFA2)
```

# Data
Modify the data directory below to point to the folder containing the downloaded data from heibox: https://heibox.uni-heidelberg.de/d/4c4457f7dc5644d4a26e/ 
```{r}
data_dir <- "data/"
```


# Processing and visualising the data
The first step in every single-cell experiment is to get the data into a usable format and to do quality control. Since these steps can often be quite resource intensive, we already provide partly preprocessed data for you. If you are interested in the preprocessing you can have a look at the [full analysis folder of the paper](https://github.com/rargelaguet/scnmt_gastrulation). You can also find links to the full data there. Here we provide you with a metadata file that will give an overview of the quality control and links samples across the different layers. 

## Metadata
```{r}
metadata <- fread(file.path(data_dir, "sample_metadata_filtered_clean.txt"))
metadata
```

Columns:
- **sample**: cell ID  
- **pass_rnaQC**: did the cell pass QC for RNA expression?.  
- **pass_metQC**: did the cell pass QC for DNA methylation? `NA` if the cell was only profiled for RNA.  
- **pass_accQC**: did the cell pass QC for chromatin accessibility? `NA` if the cell was only profiled for RNA. 
- **stage**: developmental stage.  
- **lineage**: cell type annotation (derived from mapping the cells to the [10x mouse gastrulation reference atlas](https://www.nature.com/articles/s41586-019-0933-9)).

## RNA data
In this markdown, we will focus on the mRNA expression data and how to do various processing steps with Seurat.

```{r}
# read in RNA Seurat object
rna <- readRDS(file.path(data_dir, "rna/seurat_object.rds"))
rna
```

**Tasks:**
1.) Add the metadata we read in above to the Seurat object 
2.) Add the total number of counts per cell and the number of expressed genes to the metadata of the Seurat object and plot their distribution in violin plots
3.) Normalize the data (count depth scaling and log-transformation)
4.) Find the top 1000 most highly variable genes
5.) Scale the data to mean 0 and unit variance
6.) Run and plot a PCA

Hint: The usage of most Seurat functions needed for these tasks is shown in this tutorial: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html


**Task: What is captured by the first principal components? Do you see a separation of the different stages and lineages in a UMAP plot? What can you conclude from these plots? **



# Optional: Integrating the data from different omic layers using MOFA

In case, there is time left you can have a look at a pre-trained MOFA model on this single-cell data set which combines the information from all three omic layers. Here we skip the model training part that you have seen yesterday and focus on the downstream characterisation of a pretrained MOFA model. If you are interested in the code to train the model from the data linked above, have a look at the script `train_MOFA_model.R` in the data folder.

## Load pre-computed MOFA model

As input to the model DNA methylation and chromatin accessibility values are quantified over two different sets of regulatory elements: gene promoters and enhancer elements. RNA expression was quantified over protein-coding genes. After data processing, separate views were defined for the RNA expression and for each combination of genomic context and epigenetic readout.

Note that for this tutorial we selected the MOFA Factors that explained at least 1% of variation in the RNA expression.
```{r}
MOFAobject <- readRDS(file.path(data_dir, "/MOFAmodel.rds"))
MOFAobject
```
**Task: Explore the print-out of the MOFAobject. How many cells were used for training the model?** 

We use the `samples_metadata` function on the MOFAobject to explore the meta data of the cells. The columns refer to
- **sample**: cell ID  
- **stage**: developmental stage.  
- **lineage**: cell type annotation (derived from mapping the cells to the [10x reference atlas](https://www.nature.com/articles/s41586-019-0933-9)).  
- **pass_rnaQC**: did the cell pass QC for RNA expression?.  
- **pass_metQC**: did the cell pass QC for DNA methylation? `NA` if the cell was only profiled for RNA.  
- **pass_accQC**: did the cell pass QC for chromatin accessibility? `NA` if the cell was only profiled for RNA.  
- **group**: ignore this column

```{r}
head(samples_metadata(MOFAobject))
```

**Task:  How many cells from each lineages are in the object?**




Notice that there a lot of cells that only have RNA expression measurements. One of the major advantages of MOFA is that it handles missing values, so we don't have to remove these cells prior to model training

**Task: How many cells do not have DNA methylation measurements?**

Remember that those cells have `NA` in the `pass_metQC` column.



## Overview of training data
The function `plot_data_overview` can be used to obtain an overview of the input data. 
It shows how many views (rows) and how many cells (columns) exist, what are their corresponding dimensionalities are. It also shows which views each cell is missing.
```{r fig.align="center"}
view.colors <- c(
  "RNA" = "#3CB54E",
  "Enhancers accessibility" = "#00BFC4",
  "Promoters accessibility" = "#00BFC4",
  "Enhancers methylation" = "#F37A71",
  "Promoters methylation" = "#F37A71"
)
view.colors = view.colors[views_names(MOFAobject)]

plot_data_overview(MOFAobject, colors = view.colors)
```

The data in the MOFA model has been normalised before training. For RNA data this is similar to what you have done in the `Seurat` workflow above (here a slighlty different approach was used called `scran`), for the epigenetic data m-values were used. MOFA uses a normal likelihood to model these resulting values in all modalitites

**Task: Plot a histogram of the values for the RNA, methylation and accessibility views. How good is the normal distribution as an approximation for the data? Optional: What could be other ways of modelling this distribution? **

Hint: You can extract the values using the `get_data` function. For details on the function call `?get_data`. As the results will be a nested list you have to extract the matrix from it for plotting, e.g. `get_data(MOFAobject, views="RNA")[[1]][[1]]`


## Variance decomposition analysis
The most important insight that MOFA generates is the variance decomposition analysis using `plot_variance_explained`. This plot shows the percentage of variance explained by each factor in each data modality. From this we can identify shared sources of variation across multiple omic modalities.

```{r}
plot_variance_explained(MOFAobject) +
  theme(
    axis.text.x = element_text(angle=25, hjust=1, vjust=1.05)
  )
```

**Task: What insights from the data can we learn just from inspecting this plot?**



## Characterisation of Factors

There are a few systematic strategies to characterise the molecular signal that underlies each MOFA factor:

- **Association analysis between the sample metadata and the Factor values**: function `correlate_factors_with_covariates`
- **Inspection of factor values**: functions `plot_factor` (one factor at a time) and `plot_factors` (combinations of factors)
- **Inspection of the feature weights**: functions `plot_weights` (all weights), `plot_top_weights` (only the top weights)
- **Gene set enrichment analysis on the mRNA weights**: functions `run_enrichment`, followed by `plot_enrichment`.

## Characterisation of Factor 1

### Factor values 

Plotting Factor 1 values and colouring cells by lineage assignment shows that this factor captures the variation that is associated with the separation between Mesoderm (positive Factor values) and non-Mesoderm cells (negative Factor values).
 
```{r}
plot_factor(MOFAobject,
  factor = 1,
  color_by = "lineage", 
  add_violin = TRUE,
  dodge = TRUE
) 
```

*How do we interpret the factor values?*
Each factor captures a different source of variability in the data. Mathematically, each Factor is defined by a linear combination of the input features. Each Factor ordinates cells along a one-dimensional axis that is centered at zero. Samples with different signs manifest opposite phenotypes along the inferred axis of variation, with higher absolute value indicating a stronger effect.  
Note that the interpretation of MOFA factors is analogous to the interpretation of the principal components in PCA.


### RNA weights

The weights provide a score for each gene on each factor. Genes with no association with the factor are expected to have values close to zero, whereas genes with strong association with the factor are expected to have large absolute values. The sign of the weight indicates the direction of the effect: a positive weight indicates that the feature is more active in the cells with positive factor values, and viceversa.  

Let's plot the distribution of weights for Factor 1.
```{r, warnings=FALSE, message=FALSE}
plot_weights(MOFAobject,
  view = "RNA",
  factor = 1,
  nfeatures = 10,     # Top number of features to highlight
  scale = T           # Scale weights from -1 to 1
)
```

If you are not interested in the full distribution, but just on the top weights, you can instead do:
```{r}
plot_top_weights(MOFAobject, 
  view = "RNA", 
  factor = 1, 
  nfeatures = 10,
  scale = T, 
  abs = T
)
```

We expect that genes with large positive weights For Factor 1 to be highlighy expressed in the Mesoderm cells. If we plot Factor 1 colouring cells by gene expresion of the top genes **with positive weight**:
```{r}
genes <- c("Phlda2","Mesp1")

for (i in genes) {
  plot_factor(MOFAobject,
    factor = 1,
    dot_size = 2.5,
    group_by = "lineage",
    color_by = i
  ) %>% print
}
```


Similarly, we expect that genes with large negative weights For Factor 1 to be lowly expressed in the Mesoderm cells. If we plot Factor 1 colouring cells by gene expresion of the top genes **with negative weight**:
```{r}
genes <- c("Cldn6","Pim2")

for (i in genes) {
  plot_factor(MOFAobject,
    factor = 1,
    dot_size = 2.5,
    group_by = "lineage",
    color_by = i
  ) %>% print
}
```

### RNA expression patterns in the high-dimensional space

The weights are useful to identify which genes are driving each factors. After inspecting the weights it is good practice to go back to the high-dimensional space and check if the variability that MOFA captures is real.  
For example, one could generate a heatmap plot of the RNA expression for the top genes, where samples are sorted by the corresponding factor values. This is the aim of the `plot_data_heatmap` function:

```{r}
plot_data_heatmap(MOFAobject, 
  view = "RNA", 
  factor = 1, 
  features = 25,
  annotation_samples = "lineage",
  # extra arguments passed to `pheatmap`,
  show_colnames = F, cluster_cols = F
)
```

An interesting option of `plot_data_heatmap` is to plot "denoised" observations. This is obtained by reconstructing the data using the matrix factorisation equation from MOFA:  

$$\hat{\mathbf{Y}}^m = \mathbf{W}^m\mathbf{Z}$$
where $\mathbf{W}^m$ is the weight matrix for the $m$-th view, and $\mathbf{Z}$ is the (shared) factor matrix.  
This data reconstruction step essentially removes all the variation that is not captured by the model:

```{r}
plot_data_heatmap(MOFAobject, 
  view = "RNA", 
  factor = 1, 
  denoise = TRUE,
  features = 25,
  # extra arguments passed to `pheatmap`
  show_colnames = F, cluster_cols = F 
)
```

### DNA methylation weights

As we have done with RNA, we can also visualise the distribution of weights for the epigenetic modalities. The problem about this is that the large majority of enhancers are not well annotated and we only have the genomic coordinates for them...

```{r}
plot_weights(MOFAobject,
  view = c("Enhancers methylation"),
  factor = 1,
  nfeatures = 5,
  scale = F
)
```

### DNA methylation patterns in the high-dimensional space

As done with the RNA above, let's visualise in the high-dimensional space the DNA methylation variation that MOFA captures using the `plot_data_heatmap` function. Notice how noisy and sparse DNA methylation data is. 

```{r, out.width="130%"}
plot_data_heatmap(MOFAobject, 
  view = "Enhancers methylation", 
  factor = 1, 
  features = 25,
  annotation_samples = "lineage",
  # extra arguments passed to `pheatmap`
  show_colnames = F, cluster_cols = F, 
  fontsize = 6
)
```

We will use MOFA to impute the missing values. This is based on the data reconstruction equation shown above.

```{r}
MOFAobject <- impute(MOFAobject)
```

Plot heatmap with `impute=TRUE` argument. 
```{r, out.width="130%"}
plot_data_heatmap(MOFAobject, 
  view = "Enhancers methylation", 
  factor = 1, 
  impute = TRUE,
  features = 25,
  annotation_samples = "lineage",
  # extra arguments passed to `pheatmap`
  show_colnames = F, cluster_cols = F, 
  fontsize = 6
)
```

As we guessed from the variance decomposition analysis, the promoters do not display interesting signal during germ layer commitment
```{r, out.width="130%"}
plot_data_heatmap(MOFAobject, 
  view = "Promoters methylation", 
  factor = 1, 
  impute = TRUE,
  features = 25,
  annotation_samples = "lineage",
  # extra arguments passed to `pheatmap`
  show_colnames = F, cluster_cols = F, 
  fontsize = 6
)
```


## Characterisation of Factor 2

**Task: Can you use the functions and analyses strategies introduced above to characterize Factor 2 in the model? **

- What germ layer does it capture? 
- Which genes' expression is linked to it? 
- How do methylation values change along this facotr?


## 2-dimensional representation
The first two factors capture in two dimensions all the variation that is required to separate the three germ layers. It corresponds to [Figure 2b in the original paper](https://www.nature.com/articles/s41586-019-1825-8).  

```{r}
plot_factors(MOFAobject,
  factors = c(1,2), 
  color_by = "lineage",
  dot_size = 2,
  legend = TRUE
) 
```
**Task: Change the `color_by` argument to a gene name to colour the scatterplot above by the expression of genes that you have identified in the previous analyses as markers for the the different germ layers with highest or lowest weight on the factor. Is their expression in line with the variation captures in the 2-dimensional plot?**


# sessionInfo
```{r}
sessionInfo()
```

# Contributors
Project material is based on material by Britta Velten, Ricard Argelaguet and Max Frank.
